### PR TITLE
Exposes laws as a public library

### PR DESCRIPTION
**Pull request by gr-im**
I've taken up @xvw's work to expose the validation/verification of laws as a library. As currently, `Stdlib` is still in `Preface` (waiting for `Postface`), I have kept QCheck under the `Preface` umbrella. But when we release the standard "extracted from Preface" library, we can have a `preface_qcheck` dependent on `qcheck`.

The code isn't very interesting, but at least the laws are liberalised and I've tried to cover the missing cases (`Divisible` and `Decidable`) and the handling of contravariants (and profonors/arrows) is now standardised.

You may realise that the sequence of laws (e.g. `Functor -> Apply -> Applicative -> Selective`) may make some tests redundant (i.e. `Functor` tested at all stages) but as anyone can build, for example, a `Selective` by the path they want, I didn't want to complicate the API (and that's OK for redundancy!)

### CODE REVIEW

File: lib/preface_laws/preface_laws.mli [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+module Semigroupoid = Semigroupoid
+module Category = Category
+module Arrow = Arrow
+module Arrow_zero = Arrow_zero
+module Arrow_alt = Arrow_alt
+module Arrow_plus = Arrow_plus
+module Arrow_choice = Arrow_choice
+module Arrow_apply = Arrow_apply
+
```
Comment (xhtmlboi) [inline lib/preface_laws/preface_laws.mli:59]: I think you could also generalize the laws for the different transformers (monadic and comonadic), but you can leave that to another RP.

File: lib/preface_qcheck/model.mli [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    types (notably the monoid hierarchy). *)
+
+module type COVARIANT_0 = sig
+  type t
+
+  val generator : t QCheck2.Gen.t
+  val observable : t QCheck2.Observable.t
+  val pp : Format.formatter -> t -> unit
+  val equal : t -> t -> bool
+end
```
Comment (xhtmlboi) [inline lib/preface_qcheck/model.mli:32]: `T0` and `COVARIANT_0` seems equal. Why not `module COVARIANT_0 = T0`?

File: test/preface_laws_test/equivalence.ml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+module Contravariant_suite =
```
Comment (xhtmlboi) [inline test/preface_laws_test/equivalence.ml:1]: Invariant suite seems missing ;)

File: test/preface_laws_test/predicate.ml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+module Contravariant_suite =
```
Comment (xhtmlboi) [inline test/preface_laws_test/predicate.ml:1]: Like fo `Equivalence`, `Invariant` is missing

File: test/preface_laws_test/preface_laws_test.ml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+let run count =
+  Alcotest.run "Preface Laws"
+    ( Misc.cases ~count
+    @ Identity.cases ~count
```
Comment (xhtmlboi) [inline test/preface_laws_test/preface_laws_test.ml:4]: WDYT about Adding `Approximation`  (for given Monoids, like those that are present in `Misc`?)

File: test/preface_laws_test/req.ml [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+    let hash x =
+      let a = x.Preface.Continuation.run (fun x -> x) in
+      hash obs a
+    in
+    let eq = cont_equal (equal obs) in
+    make ~eq ~hash print
+  ;;
+end
+
+module Mini_yocaml = struct
```
Comment (xhtmlboi) [inline test/preface_laws_test/req.ml:334]: Very cute <3

File: test/preface_laws_test/try.ml [hunk 7]
```diff
@@ -0,0 +1,79 @@
+module Invariant_suite =
+  Preface.Qcheck.Invariant.Suite (Req.Try) (Preface.Try.Invariant) (Sample.Int)
+    (Sample.String)
+    (Sample.Float)
+
+module Functor_suite =
+  Preface.Qcheck.Functor.Suite (Req.Try) (Preface.Try.Functor) (Sample.Int)
+    (Sample.String)
+    (Sample.Float)
+
```
Comment (xhtmlboi) [inline test/preface_laws_test/try.ml:10]: `Alt` is missing

File: test/preface_laws_test/validate.ml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    (Sample.Int)
+    (Sample.String)
+    (Sample.Float)
+
+module Functor_suite =
+  Preface.Qcheck.Functor.Suite (Req.Validate) (Preface.Validate.Functor)
+    (Sample.Int)
+    (Sample.String)
+    (Sample.Float)
+
```
Comment (xhtmlboi) [inline test/preface_laws_test/validate.ml:12]: `Alt` is missing

**Review by xhtmlboi — CHANGES_REQUESTED** [review]
I noticed some issues. If you don't have the faith to deal with approximations and transformers in this branch, no worries!

File: lib/preface_laws/preface_laws.mli (hunk 1 above)
Comment (gr-im) [reply lib/preface_laws/preface_laws.mli:59]: I've added Writer, but `State` and `Reader` (and the comonads) need a bit more work so I suggest leaving that for a later one.

Comment (xhtmlboi) [reply lib/preface_laws/preface_laws.mli:59]: Yeah do not worry!

**Review by xhtmlboi — APPROVED** [review]

**Review by xvw — APPROVED** [review]
Damn, thanks for taking over this big job!

**Review by d-plaindoux — APPROVED** [review]

**Pull request merged**